lib/Partitioner/Partitioner.cpp

```diff
@@ -0,0 +1,365 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Partitioner/Partitioner.h"
+#include "glow/Graph/Utils.h"
+
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace glow;
+using llvm::isa;
+
+/// Margin for code itself -- should be adjusted later.
+#define MARGIN 1024 * 128
+
+struct BFSLevel {
+  /// A list of <level, nodelist> with BFS order.
+  std::vector<std::pair<int, std::vector<Node *>>> levels;
+  /// A set of visited nodes.
+  std::unordered_set<const Node *> visited;
+};
+
+Partitioner::Partitioner(Module *parent, std::vector<DeviceInfo> &devices)
+    : deviceInfo_(devices) {
+  module_ = parent;
+  // Check if all the functions are in the same structure.
+  if (!isSameFunctionFamily()) {
+    assert("not in the same function family");
+  }
+  memSize = module_->getConstantsSize();
+}
+
+/// Get the represent function : the one with the largest input, and update the
+/// memSize.
+void Partitioner::selectRepFunc() {
+  auto funcList = module_->getFunctions();
+  for (Function *F : funcList) {
+    size_t size = memSize;
+
+    // The set to keep the placeholders (only for Inputs) whose size is already
+    // calcualted.
+    std::set<std::string> pSet;
+
+    for (auto &node : F->getNodes()) {
+      int n = node.getNumInputs();
+      if (node.getKind() == Kinded::Kind::SaveNodeKind) {
+        // Special node, the placehodler should be ignored?
+        continue;
+      }
+      for (int i = 0; i < n; i++) {
+        Placeholder *in =
+            llvm::dyn_cast<Placeholder>(node.getNthInput(i).getNode());
+        if (in && (pSet.find(in->getName()) == pSet.end())) {
+          auto ty = in->getType();
+          size += ty->getSizeInBytes();
+          pSet.insert(in->getName());
+        }
+      }
+    }
+    if (size > memSize) {
+      F_ = F;
+      memSize = size;
+    }
+  }
+}
+
+/// Get the minimal memory requirement (constant) for each op in the function.
+void Partitioner::initOpMemUsage() {
+  memUsage.clear();
+  for (auto &node : F_->getNodes()) {
+    int n = node.getNumInputs();
+    unsigned size = 0;
+    if (node.getKind() != Kinded::Kind::SaveNodeKind) {
+      for (int i = 0; i < n; i++) {
+        Storage *in = llvm::dyn_cast<Storage>(node.getNthInput(i).getNode());
+        if (in) {
+          auto ty = in->getType();
+          size += ty->getSizeInBytes();
+        }
+      }
+    }
+    memUsage[&node] = size;
+  }
+}
+
+static BFSLevel getBFSLevel(Function *F) {
+  // Visit graph nodes in BFS order. For each non-storage node, get its level.
+  // Use the preorder to get the roots. Now assume there is only one output op
+  // (i.e. root) now.
+  GraphPreOrderVisitor visitor(*F);
+  Node *node = nullptr;
+  for (auto &N : visitor.getPreOrder()) {
+    if (isa<Storage>(N))
+      continue;
+    node = N;
+    break;
```

Comment:
I'm confused, won't this just quit at the first non-storage node? Why can't there be multiple in the first level, is this the save node?

Reply:
This is a save node.  The comment mentioned that now we assume there is only one root now. The current models I've met only have 1 output node. But it could be multiple roots as well later.